Refactor ServiceManager/ServiceBroker

notspiff opened the pull request:
This is the start of my service container refactor. see first commit for why I believe this approach is an improvement.

It's early, feel free to review but some things you can skip
- yes, i'm initing services directly in Application right now. I wil change this in the end, but I want to see how things work out first. Ultimately, I hope to make this a map of types registered in one spot, and then have a separate call to bring services up. Bu to do this we need to change constructors to not do stuff before StartService is called so right now I do not try to factor it nicely into a function.

FernetMenta commented on xbmc/services/ServiceManager.cpp:
```diff
@@ -0,0 +1,49 @@
+  size_t ok = 0;
+  for (auto it = m_services[level].rbegin(); it != m_services[level].rend(); ++it, ++ok)
+  {
+    if (!(*it)->StopService())
```
the lock pretends that stopping the service here is secure but it is not. it stops the service while a client may hold a pointer to an interface.

Since you dropped the logic for bringing up the serviecs from this class, you can drop start/stop completely. An implicit call to StartService in RegisterService seems not right anyway.

Stopping services is not synchronous. We should send signals, requestStop, to the registered services and wait for acc. 
In addition ServiceMangager could track if it handed away interfaces. Then only request stop if the service is not used by any clients.

notspiff replied:
It is not pretending anything. It protects m_services. Ref counting takes care of any holders and it is the service' responsibity to handle (ignore) calls after stop. I thought i was verbose enough in pr desc; the start/stop thing incl making it async will be handled properly later but it takes a lot of additional refactoring and i am not ready to target that yet.

FernetMenta replied:
Controlling and visualizing dependencies between services was the idea for current service mamanger. By taking this away there is not much left that justifies the name "manager".  Maybe I miss more things you have planned with this refactor. But maybe ServceRegistry is a better fit for what it does.

notspiff replied:
??? It is a 100% direct replacement for the current code except without all the boilerplating.

FernetMenta replied:
Current ServiceManager has Init method(s) that starts the services in required order. By looking at it, you get an idea of how services depend on each other. I named it manager becuase it was supposed to manage the lifetime of the services and gain control over the sequence how they need to be started.
You said that you moved this part into CApp but it's not clear to me where you intend to move it finally.

wsnipex commented on xbmc/Application.cpp:
```diff
@@ -402,6 +404,24 @@ bool CApplication::Create(const CAppParamParser &params)
+#elif defined(HAS_WIN10_NETWORK)
+  SERVICES::CServiceManager::GetInstance().RegisterService(std::shared_ptr<CNetwork>(new CNetworWin10()));
+#else
+  SERVICES::CServiceManager::GetInstance().RegisterService(std::shared_ptr<CNetwork>(new CNetwor()));
```
CNetwork
same above for win32 and win10

xhaggi commented on xbmc/Application.cpp:
```diff
@@ -398,6 +399,7 @@ bool CApplication::Create(const CAppParamParser &params)
 #ifdef HAS_PYTHON
   SERVICES::CServiceManager::GetInstance().RegisterService(std::shared_ptr<XBPython>(new XBPython), 0);
 #endif
+  SERVICES::CServiceManager::GetInstance().RegisterService(std::shared_ptr<PLAYLIST::CPlayListPlayer>(new PLAYLIST::CPlayListPlayer));
```
should we use `std::make_shared` instead where possible?

notspiff replied:
sure.

notspiff committed:
changed: move CNetwork into new ServiceManager

notspiff committed:
changed: move XBPython into new ServiceManager

notspiff committed:
changed: move AnnouncementManager into new ServiceManager

notspiff committed:
changed: move PlaylistPlayer into new ServiceManager

notspiff committed:
changed: rename CNetwork to CNetworkBase

and add typedefs to CNetwork for specific platforms.
in preparation for rtti'ing the service

notspiff committed:
add a new service manager

this replaces the current service manager/broker.

advantages:
  - no need for explicitly coding all the members in two classes - just
    add and be done
  - no need for the excessive forwarding
  - no need for two layers of accessors - use rtti
  - no blind deref'ing of ptrs - return ptrs to user. this allows
    running without services up
  - proper locking

notspiff commented:
obviously not happening like this.
